### DISCUSSION TITLE
Can XState do stateless conditional actions?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [statelyai/xstate#1790] [Q&A] [answered]
I've read about `actions.choose`, but it seems to not serialize at all into the viz tool.

I have some networking code that looks like:
```ts
    switch (decrypted.type) {
      case 'JOIN':
        if (this.isKnown(data.uid)) {
          if (this.currentGame) {
            this.sendStateTo(data.uid);
          } else {
            this.__updatePlayersWithParticipants();
          }
        } else {
          this._addPlayer(decrypted, data.uid);
        }

        return;
      case 'SYN':
        if (this.currentGame) {
          this.ifKnown(data.uid, () => this.sendToHex({ type: 'ACK' }, data.uid));

          return;
        }

        this.sendToHex({ type: 'ACK' }, data.uid);

        return;
```

and I'd like to represent this as a statechart, but the only way I've found to do conditional actions is via `actions.choose`. Is there another way? will actions.choose be supported by the visualizer eventually?

Not every if statement can or should be its own state.
For example, when receiving a `JOIN` message, there will be two states, `waiting-for-players` and `game-active` and there are different behaviors to do if we know of the player already or not

Also, since I'm doing networking code, is there a beta version of xstate I can use that allows sequential actions?

**Comment by davidkpiano** [comment, marked as answer]
> will `actions.choose` be supported by the visualizer eventually?

Yes, it will. I would use `actions.choose(...)` and then wait for a visualizer update to magically show that data 🔜 

> Also, since I'm doing networking code, is there a beta version of xstate I can use that allows sequential actions?

What do you mean by sequential actions?

**Reply by NullVoxPopuli** [reply]
> What do you mean by sequential actions?

like, if I want to do:

- addPlayer
- broadcastPlayerJoined (to all other players)

the broadcastPlayerJoined action would be dependent on data created from addPlayer (where some data is managed outside the machine)

**Reply by davidkpiano** [reply]
Seems like you can do this either in an action (if fire-and-forget):

```js
entry: async () => {
  const data = await addPlayer();
  await broadcastPlayerJoined(data.player);
}
```

Or, if it does need to communicate back to the machine, `invoke` instead:
```js
invoke: {
  src: () => async () => {
    const data = await addPlayer();
    const result = await broadcastPlayerJoined(data.player);
    return result;
  },
  onDone: {
    actions: (_, event) => {/* do something with event.data (result) */}
  }
}
```

EDIT: note that you won't get the benefits of having each sequential step fully controlled/visualized as a state machine unless you make explicit states for each. This is the "easiest" approach.

**Reply by a-y-u-s-h** [reply]
I think @NullVoxPopuli wants to know whether we can have name-less states to execute actions sequentially or even better, execute smaller actions specified in an object outside of state representation one after another instead of naming each transition state, like just passing multiple actions in an array and they execute one after the other instead of executing parallely like they do now (or maybe another key could be added with actions specifying whether they should be executed in parallel or sequentially?). I would personally like this feature as well if possible, because sometimes it's a pain to merge actions into one service and then name that transition state when all I want is some actions to run one after another in the order I specify them in the array.

**Reply by Andarist** [reply]
Actions execute sequentially - `assign` is currently special-cased though and is executed first. This has already been corrected on the `next` branch but it has to wait till the next major version before we can release it.

**Reply by a-y-u-s-h** [reply]
I think that should be documented as well. Because I remember @davidkpiano  once mentioned that in a talk somewhere that you should not assume that actions in the array happen after the previous one ends but rather they're independent of each other and if the order of actions is important for a program then you should create a state in between them. What if I have multiple `assign` actions in the array? Do all assign actions happen sequentially as they're mentioned as well?

**Reply by Andarist** [reply]
>  think that should be documented as well. Because I remember @davidkpiano once mentioned that in a talk somewhere that you should not assume that actions in the array happen after the previous one ends but rather they're independent of each other and if the order of actions is important for a program then you should create a state in between them.

IMHO guaranteeing the sequence is reasonable - after all, we have to call them in some order and the "document order" is a reasonable one. It is also "required" by the SCXML spec - to which we try to adhere. 

> What if I have multiple assign actions in the array? Do all assign actions happen sequentially as they're mentioned as well?

Yes.

**Reply by davidkpiano** [reply]
> Because I remember @davidkpiano once mentioned that in a talk somewhere that you should not assume that actions in the array happen after the previous one ends but rather they're independent of each other and if the order of actions is important for a program then you should create a state in between them

Yes, the sequence order should be guaranteed (and should be by document order as @Andarist mentioned). What I meant is that you should not rely on _side effects_ from other actions; that is, if you care about an action's outcome, it should not be a fire-and-forget action. It should be something else, like an invocation.
